### PR TITLE
Finalize Bitcoin refactor

### PR DESCRIPTION
**Pull request by tsusanka**
Implements most of subissues in https://github.com/trezor/trezor-firmware/issues/617.

I will open another issue for mypy, because that is more time-consuming unfortunately.

### CODE REVIEW

File: core/src/apps/bitcoin/sign_tx/scripts.py [hunk 1]
```diff
@@ -89,13 +86,13 @@ def output_derive_script(txo: TxOutputType, coin: CoinInfo) -> bytes:
         elif version == cashaddr.ADDRESS_TYPE_P2SH:
             version = coin.address_type_p2sh
         else:
-            raise ScriptsError("Unknown cashaddr address type")
```
Comment (tsusanka) [inline core/src/apps/bitcoin/sign_tx/scripts.py:92]: I believe this was a bug (missing FailureType). Occurs in other places in scripts.py and decred.py.

File: core/src/apps/bitcoin/get_address.py [hunk 2]
```diff
@@ -2,11 +2,12 @@
 from trezor.messages import InputScriptType
 from trezor.messages.Address import Address
 
+from .sign_tx import addresses
```
Comment (matejcik) [inline core/src/apps/bitcoin/get_address.py:5]: all non-signers (here `addresses.py`, `multisig.py`) should be moved to `apps/bitcoin`, seeing as the other operations are also using them

File: core/src/apps/bitcoin/sign_tx/scripts.py (hunk 1 above)
Comment (matejcik) [reply core/src/apps/bitcoin/sign_tx/scripts.py:92]: agreed. that also shows that we don't have test coverage for this.
(maybe it doesn't really matter though?)

File: core/src/apps/common/signverify.py [hunk 3]
```diff
@@ -4,9 +4,9 @@
 from trezor.crypto.hashlib import blake256, sha256
 from trezor.ui.text import Text
 
+from apps.bitcoin.sign_tx.writers import write_varint
```
Comment (matejcik) [inline core/src/apps/common/signverify.py:7]: maybe we should move `write_varint` to `apps.common.writers`?

maybe rename it to `write_bitcoin_varint` to differentiate from protobuf varint?

File: core/src/apps/bitcoin/sign_tx/layout.py [hunk 4]
```diff
@@ -54,50 +54,39 @@ async def confirm_output(
...
 
     text = Text("High fee", ui.ICON_SEND, ui.GREEN)
     text.normal("The fee of")
     text.bold(format_coin_amount(fee, coin))
     text.normal("is unexpectedly high.", "Continue?")
-    return await confirm(ctx, text, ButtonRequestType.FeeOverThreshold)
+    await require_confirm(ctx, text, ButtonRequestType.FeeOverThreshold)
 
 
-async def confirm_foreign_address(
```
Comment (matejcik) [inline core/src/apps/bitcoin/sign_tx/layout.py:87]: was this function unused?

File: core/src/apps/bitcoin/sign_tx/scripts.py (hunk 1 above)
Comment (tsusanka) [reply core/src/apps/bitcoin/sign_tx/scripts.py:92]: Indeed. Honestly I was wondering if you will suggest to write the test yourself :). We can do it, but writing test for each failing condition seems a bit annoying and maybe indeed redundant.

File: core/src/apps/bitcoin/sign_tx/layout.py (hunk 4 above)
Comment (tsusanka) [reply core/src/apps/bitcoin/sign_tx/layout.py:87]: Yes AFAICT. We are using `paths.show_path_warning` in `__init__` instead.

File: core/src/apps/bitcoin/get_address.py (hunk 2 above)
Comment (tsusanka) [reply core/src/apps/bitcoin/get_address.py]: Fixed in 4497ea8a957a16c2bc9885a88c7831c55872b0bd

Comment (tsusanka) [reply core/src/apps/bitcoin/get_address.py]: Not really, same goes for writers.py

**Comment by tsusanka** [comment]
If you forgive me I will squash into one commit and rebase on master, it will be significantly easier because of the keychain merge.

File: core/src/apps/common/signverify.py (hunk 3 above)
Comment (tsusanka) [reply core/src/apps/common/signverify.py]: Done in e5856b3e6ed7ee092a6dc3b280ddf35d9ba605c4.

**Comment by tsusanka** [comment]
Rebased and force-pushed. It wasn't exactly an easy one :|.

Things to do:
- [ ] add tests with mixed inputs as mentioned [in the issue](https://github.com/trezor/trezor-firmware/issues/617#issuecomment-629135257)
- [ ] look if we can merge bitcoin writers with the common ones (some of it).

**Comment by tsusanka** [comment]
>  look if we can merge bitcoin writers with the common ones (some of it).

I have moved decred writers to decred.py in e5fc148. I think the rest is OK.

File: core/src/apps/bitcoin/get_address.py (hunk 2 above)
Comment (tsusanka) [reply core/src/apps/bitcoin/get_address.py]: Should be done, now.

File: tests/device_tests/test_msg_signtx_mixed_inputs.py [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the License along with this library.
+# If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.
+
+import pytest
+
+from trezorlib import btc, messages as proto
```
Comment (matejcik) [inline tests/device_tests/test_msg_signtx_mixed_inputs.py:19]: please rename this to `messages`

File: tests/device_tests/test_msg_signtx_mixed_inputs.py [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+from trezorlib.tools import parse_path
+
+from ..tx_cache import TxCache
+
+B = proto.ButtonRequestType
```
Comment (matejcik) [inline tests/device_tests/test_msg_signtx_mixed_inputs.py:24]: this seems unused

File: tests/device_tests/test_msg_signtx_mixed_inputs.py (hunk 5 above)
Comment (tsusanka) [reply tests/device_tests/test_msg_signtx_mixed_inputs.py]: Fixed.

File: tests/device_tests/test_msg_signtx_mixed_inputs.py (hunk 6 above)
Comment (tsusanka) [reply tests/device_tests/test_msg_signtx_mixed_inputs.py]: Fixed.

**Comment by tsusanka** [comment]
> it's unfortunate that the "reorder imports" thing got squashed in, that would have been nice to keep separate.

I have realized I can fix this easily by reverting the setup.cfg then regenerating, squashing and then introducing the change again :). I will give it a go.

**Pull request merged**